src/test/java/org/testng/internal/DynamicGraphTest.java

```diff
@@ -289,4 +290,59 @@ public void testDuplicationFunctionality() {
       assertThat(tla.getPassedTests().size()).isEqualTo(expected);
   }
 
+  @Test(expectedExceptions = IllegalStateException.class)
+  public void testPreventCycles() {
+    // Cycles having the same weight should throw an IllegalStateException immediately.
+    DynamicGraph<String> dg = new DynamicGraph<>();
+    dg.addEdge(0, "a", "b");
+    dg.addEdge(0, "b", "a");
+  }
+
+  @Test
+  public void testAllowedCycles() {
+    // Cycles with differing weights are explicitly allowed as the graph will discard the lowest weighted edge in
+    // order to free up the cycle.
+    DynamicGraph<String> dg = new DynamicGraph<>();
+    dg.addEdge(0, "a", "b");
+
+    // Different weight is allowed
+    dg.addEdge(1, "b", "a");
+  }
+
+  @Test
+  public void testGainWeight() {
+    DynamicGraph<String> dg = new DynamicGraph<>();
+    dg.addEdge(1, "a", "b");
+    assertThat(dg.getEdges().values().size()).isEqualTo(1);
+    assertThat(dg.getEdges().get("a").get("b")).isEqualTo(1);
+
+    // Duplicated edge, but with lower weight
+    dg.addEdge(0, "a", "b");
+
+    // Should only be one edge with same weight
+    assertThat(dg.getEdges().values().size()).isEqualTo(1);
+    assertThat(dg.getEdges().get("a").get("b")).isEqualTo(1);
+
+    // Duplicated edge, but with higher weight
+    dg.addEdge(2, "a", "b");
+
+    // Should only be one edge with weight of 2 now.
+    assertThat(dg.getEdges().values().size()).isEqualTo(1);
+    assertThat(dg.getEdges().get("a").get("b")).isEqualTo(2);
+  }
```

Comment:
The above three tests were not covered previously but are important to how DynamicGraph works.

Reply:
:+1: